src/Storages/Kafka/StorageKafka.cpp

```diff
@@ -244,69 +244,130 @@ namespace
+                        continue;
+                    loadTopicConfig(kafka_config, config, kafka_topic_path);
                 }
+                continue;
             }
+            if (tag.starts_with(CONFIG_KAFKA_TAG)) /// skip legacy configuration per topic e.g. <kafka_TOPIC_NAME>.
+                continue;
+            // Update configuration from the configuration. Example:
+            //     <kafka>
+            //         <retry_backoff_ms>250</retry_backoff_ms>
+            //         <fetch_min_bytes>100000</fetch_min_bytes>
+            //     </kafka>
+            loadConfigProperty(kafka_config, config, config_prefix, tag);
+        }
+    }
+
+    void loadTopicLegacyConfig(cppkafka::Configuration & kafka_config, const Poco::Util::AbstractConfiguration & config, const String & prefix, const Names & topics)
+    {
+        // Update consumer topic-specific configuration (legacy syntax, retained for compatibility). Example with topic "football":
+        //     <kafka_football>
+        //         <retry_backoff_ms>250</retry_backoff_ms>
+        //         <fetch_min_bytes>100000</fetch_min_bytes>
+        //     </kafka_football>
+        // The legacy syntax has the problem that periods in topic names (e.g. "sports.football") are not supported because the Poco
+        // configuration framework hierarchy is based on periods as level separators. Besides that, per-topic tags at the same level
+        // as <kafka> are ugly.
+        String topic_path = prefix;
+        if (!prefix.ends_with(CONFIG_KAFKA_TAG))
+            topic_path = prefix + "." + CONFIG_KAFKA_TAG;
+        for (const auto & topic : topics)
+        {
+            const auto topic_config_key = topic_path + "_" + topic;
+            if (config.has(topic_config_key))
+                loadFromConfig(kafka_config, config, topic_config_key, topics);
         }
     }
+
+    void loadConsumerConfig(cppkafka::Configuration & kafka_config, const Poco::Util::AbstractConfiguration & config, const String & prefix, const Names & topics)
+    {
+        const String consumer_path = prefix + "." + CONFIG_KAFKA_CONSUMER_TAG;
+        if (!config.has(consumer_path))
+            return;
+        loadFromConfig(kafka_config, config, consumer_path, topics);
+        /// Load legacy configuration per topic from the same consumer level
+        loadTopicLegacyConfig(kafka_config, config, consumer_path, topics);
+    }
+
+    void loadProducerConfig(cppkafka::Configuration & kafka_config, const Poco::Util::AbstractConfiguration & config, const String & prefix, const Names & topics)
+    {
+        const String producer_path = prefix + "." + CONFIG_KAFKA_PRODUCER_TAG;
+        if (!config.has(producer_path))
+            return;
+        loadFromConfig(kafka_config, config, producer_path, topics);
+        /// Load legacy configuration per topic from the same producer level
+        loadTopicLegacyConfig(kafka_config, config, producer_path, topics);
```

Comment:
I actually thought that we should remove the legacy syntax (basically `loadTopicLegacyConfig()`), now that it has been deprecated for a year (#46752). Perhaps that would simplify the code.

Reply:
Yes, agree! I removed it and added a warning if someone still uses it.